numpy/core/src/multiarray/compiled_base.c

```diff
@@ -1103,23 +1103,37 @@ arr_unravel_index(PyObject *self, PyObject *args, PyObject *kwds)
 
     unravel_size = PyArray_MultiplyList(dimensions.ptr, dimensions.len);
 
+    dtype = PyArray_DescrFromType(NPY_INTP);
+    if (dtype == NULL) {
+        goto fail;
+    }
+
     if (!PyArray_Check(indices0)) {
         indices = (PyArrayObject*)PyArray_FromAny(indices0,
                                                     NULL, 0, 0, 0, NULL);
         if (indices == NULL) {
             goto fail;
         }
+
+        if (PyArray_SIZE(indices) == 0) {
```

Comment:
No, that is a good point. My tests have been for the simple case of an empty list, but I originally encountered this bug when scipy's KDTree returned an empty array of indexes. Now, in that case, I am pretty sure the dtype was set properly there, but it isn't that far off to cause mayhem.

So, the question is, should I force a recast only when the array is empty, or should I do it all the time? Or, perhaps I only recast when the array is empty, and raise an error otherwise if the dtype is not compatible with intp?